### PR TITLE
Don't clone values when calculating storage root

### PR DESCRIPTION
**Pull request by bkchr**
Instead of cloning all the keys and values of the overlay when
calculating the storage root, we pass all the values by reference. This
should probably bring some performance improvements when calculating the
storage root.

### CODE REVIEW

**Comment by cheme** [comment]
no cow version https://github.com/cheme/substrate/commit/4277358ebe4cd00af0fc0331758bb6ac34457939
About the bench, I think even if there is no visible change, I am happy not having the clone.

**Comment by bkchr** [comment]
> no cow version [cheme@4277358](https://github.com/cheme/substrate/commit/4277358ebe4cd00af0fc0331758bb6ac34457939)
> About the bench, I think even if there is no visible change, I am happy not having the clone.

I added the `Cow` out of reason, because otherwise `full_storage_root` could not be implemented. Did it worked with your branch?

File: client/db/src/bench.rs [hunk 1]
```diff
@@ -193,19 +193,18 @@ impl<B: BlockT> StateBackend<HashFor<B>> for BenchmarkingState<B> {
 		}
 	}
 
-	fn storage_root<I>(&self, delta: I) -> (B::Hash, Self::Transaction) where
-		I: IntoIterator<Item=(Vec<u8>, Option<Vec<u8>>)>
-	{
+	fn storage_root<'a>(
+		&self,
+		delta: impl Iterator<Item=(std::borrow::Cow<'a, [u8]>, Option<std::borrow::Cow<'a, [u8]>>)>,
```
[the remark this answers is not in the document]
Comment (bkchr) [reply client/db/src/bench.rs:198]: It was required for `full_storage_root`. I can make `child_storage_root` use the same api.

**Comment by bkchr** [comment]
I actually started this PR because of a change I needed in Cumulus. In the end I thought that we would see some improvement of the performance, but yeah that is not right. I think to see the performance benefit we need a `MemoryDB` that uses `Cow` that does not clone the data internally.

**Pull request merged**